source/common/router/router.cc

```diff
@@ -503,20 +503,40 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
   const auto& upstream_http_protocol_options = cluster_->upstreamHttpProtocolOptions();
 
   if (upstream_http_protocol_options.has_value()) {
-    const auto parsed_authority = Http::Utility::parseAuthority(headers.getHostValue());
-    if (!parsed_authority.is_ip_address_ && upstream_http_protocol_options.value().auto_sni()) {
-      callbacks_->streamInfo().filterState()->setData(
-          Network::UpstreamServerName::key(),
-          std::make_unique<Network::UpstreamServerName>(parsed_authority.host_),
-          StreamInfo::FilterState::StateType::Mutable);
+    absl::string_view sni_value;
+    bool should_set_sni = true;
+
+    // Check whether `alt_header_name` is specified.
+    const auto alt_header_name = upstream_http_protocol_options.value().alt_header_name();
+    if (alt_header_name.empty()) {
+      // Default the SNI value to the Host/Authority header.
+      const auto parsed_authority = Http::Utility::parseAuthority(headers.getHostValue());
+      should_set_sni = !parsed_authority.is_ip_address_;
+      sni_value = std::string(parsed_authority.host_);
```

Comment:
This is unsafe. You're creating a temporary std::string, then storing a pointer to it in the string_view. You need `parsed_authority` to have a lifetime longer than the `sni_value` if you want to hold a `string_view` pointing to it.

Reply:
Drive by comment: @ggreenway Using `std::string sni_value` might be easier. The string copy is inevitable unless we scope `parse_authority` outside the branch.

Reply:
Yep, I agree. Just saying that it cannot be the way it is, and explaining why.

Reply:
you don't need creating `std::string`, the `parsed_authority.host_` points to headers so it outlives.